core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java

```diff
@@ -0,0 +1,197 @@
+        MIN_OUTPUT_FILES,
+        MIN_OUTPUT_FILES_DEFAULT);
+
+    minInputFiles = PropertyUtil.propertyAsInt(options,
+        MIN_INPUT_FILES,
+        MIN_INPUT_FILES_DEFAULT);
+
+    validateOptions();
+    return this;
+  }
+
+  @Override
+  public Iterable<FileScanTask> selectFilesToRewrite(Iterable<FileScanTask> dataFiles) {
+    return FluentIterable.from(dataFiles)
+        .filter(scanTask -> scanTask.length() < minFileSize || scanTask.length() > maxFileSize);
+  }
+
+  @Override
+  public Iterable<List<FileScanTask>> planFileGroups(Iterable<FileScanTask> dataFiles) {
+    ListPacker<FileScanTask> packer = new BinPacking.ListPacker<>(maxGroupSize, 1, false);
+    List<List<FileScanTask>> potentialGroups = packer.pack(dataFiles, FileScanTask::length);
+    return potentialGroups.stream().filter(group ->
+        numOutputFiles(group) >= minOutputFiles && group.size() >= minInputFiles
+    ).collect(Collectors.toList());
+  }
+
+  private long numOutputFiles(List<FileScanTask> group) {
+    long groupSize = group.stream().mapToLong(FileScanTask::length).sum();
+    return (long) Math.ceil((double) groupSize / targetFileSize);
+  }
+
+  private void validateOptions() {
+    if (minFileSize >= maxFileSize) {
+      throw new IllegalArgumentException(
+          String.format("Cannot use bin pack when %s is greater than %s, %d >= %d",
+              MIN_FILE_SIZE_BYTES, MAX_FILE_SIZE_BYTES, minFileSize, maxFileSize));
+    }
+    if (minFileSize > targetFileSize) {
+      throw new IllegalArgumentException(
+          String.format("Cannot use bin pack when %s is greater than %s, " +
+                  "all files written will be smaller than the threshold, %d >= %d",
+              MIN_FILE_SIZE_BYTES, RewriteDataFiles.TARGET_FILE_SIZE_BYTES, minFileSize, targetFileSize));
+    }
+    if (maxFileSize < targetFileSize) {
+      throw new IllegalArgumentException(
+          String.format("Cannot use bin pack when %s is greater than %s, " +
+                  "all files written will be larger than the threshold, %d >= %d",
+              MAX_FILE_SIZE_BYTES, RewriteDataFiles.TARGET_FILE_SIZE_BYTES, maxFileSize, targetFileSize));
+    }
+    if (minInputFiles < 1) {
+      throw new IllegalArgumentException(
+          String.format("Cannot use bin pack when %s is less than 1. All values less than 1" +
+                  "have the same effect as 1. %d < 1",
+              MIN_INPUT_FILES, minInputFiles)
+      );
+    }
+    if (minOutputFiles < 1) {
+      throw new IllegalArgumentException(
+          String.format("Cannot use bin pack when %s is less than 1. All values less than 1" +
+                  "have the same effect as 1. %d < 1",
+              MIN_OUTPUT_FILES, minOutputFiles)
+      );
+    }
+    if (targetFileSize <= 0) {
+      throw new IllegalArgumentException(
+          String.format("Cannot use a non-positive number for %s. %d <= 0",
+              RewriteDataFiles.TARGET_FILE_SIZE_BYTES, targetFileSize)
```

Comment:
nit: just `TARGET_FILE_SIZE_BYTES`?

Reply:
It's from the Action not the Strategy, I can do the static import?

Reply:
I missed it is from another class. That's fine then.